vpr/src/place/net_cost_handler.cpp

```diff
@@ -1622,9 +1725,149 @@ double NetCostHandler::get_total_wirelength_estimate() const {
+                double expected_y_wl = expected_wirelength - expected_x_wl;
+
+                int total_channel_segments = distance_x * distance_y * distance_z;
+                double expected_per_x_segment_wl = expected_x_wl / total_channel_segments;
+                double expected_per_y_segment_wl = expected_y_wl / total_channel_segments;
+
+                for (int layer = bb.layer_min; layer <= bb.layer_max; layer++) {
+                    for (int x = bb.xmin; x <= bb.xmax; x++) {
+                        for (int y = bb.ymin; y <= bb.ymax; y++) {
+                            chanx_util_[layer][x][y] += expected_per_x_segment_wl;
+                            chany_util_[layer][x][y] += expected_per_y_segment_wl;
+                        }
+                    }
+                }
+            } else {
+                const std::vector<t_2D_bb>& bb = layer_bb_coords_[net_id];
+                const vtr::NdMatrixProxy<int, 1> net_layer_pin_sink_count = num_sink_pin_layer_[size_t(net_id)];
+
+                for (size_t layer = 0; layer < bb.size(); layer++) {
+                    if (net_layer_pin_sink_count[layer] == 0) {
+                        continue;
+                    }
+
+                    double crossing = wirelength_crossing_count(net_layer_pin_sink_count[layer] + 1);
+                    double expected_wirelength = ((bb[layer].xmax - bb[layer].xmin + 1) + (bb[layer].ymax - bb[layer].ymin + 1)) * crossing;
+
+                    int distance_x = bb[layer].xmax - bb[layer].xmin + 1;
+                    int distance_y = bb[layer].ymax - bb[layer].ymin + 1;
+
+                    double expected_x_wl = (double)distance_x / (distance_x + distance_y) * expected_wirelength;
+                    double expected_y_wl = expected_wirelength - expected_x_wl;
+
+                    int total_channel_segments = distance_x * distance_y;
+                    double expected_per_x_segment_wl = expected_x_wl / total_channel_segments;
+                    double expected_per_y_segment_wl = expected_y_wl / total_channel_segments;
+
+                    for (int x = bb[layer].xmin; x <= bb[layer].xmax; x++) {
+                        for (int y = bb[layer].ymin; y <= bb[layer].ymax; y++) {
+                            chanx_util_[layer][x][y] += expected_per_x_segment_wl;
+                            chany_util_[layer][x][y] += expected_per_y_segment_wl;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Channel width is computed only once and reused in later calls.
+    if (chanx_width_.empty()) {
+        VTR_ASSERT(chany_width_.empty());
+        std::tie(chanx_width_, chany_width_) = calculate_channel_width();
+    }
+
+    VTR_ASSERT(chanx_util_.size() == chany_util_.size());
+    VTR_ASSERT(chanx_util_.size() == chanx_width_.size());
+    VTR_ASSERT(chany_util_.size() == chany_width_.size());
+
+    for (size_t layer = 0; layer < num_layers; ++layer) {
```

Comment:
Comment why you do things this way ... explain that some channels may not exist?